### PR TITLE
Generic Stack Component

### LINKED ISSUE #44: Generic Stack Component
We reuse our `Stack` component across projects by now, but it would be great to have the same implementation (at least in new projects) everywhere.

**Acceptance Criteria**
- the breakpoints can be specified via css variables (`--size-breakpoint-tablet`, `--size-breakpoint-desktop`, `--size-breakpoint-widescreen`)
  - the default (therefore no breakpoint name for it) is _mobile_ (mobile first)
  - the breakpoints would be part of the project, not the component
- make sure all developers agree on the shaping document before implementation
- there's a css file in the npm module that can be manually required (e.g. by sprockets or vite-rails) that brings in the `stack` class
- there's a `Stack` react component that can be required (`import {Stack} from '@nerdgeschoss/shimmer/components/stack'`) which exposes all the settings as typed properties

### PR DESCRIPTION
**Pull request by grcldq**
fixes #44

### CODE REVIEW

**Comment by grcldq** [comment]
- [ ] Update readme file

File: README.md [hunk 1]
```diff
@@ -1,8 +1,26 @@
...
 
 Shimmer is a collection of Rails extensions that bring advanced UI features into your app and make your life easier as a developer.
 
-## Features
+Features
+--------
+
+### SCSS Components
+
+Shimmer includes a suite of SCSS components aimed at standardizing your project's style and making your SCSS more maintainable. Two key components of this suite are the Stack Component and the Variable component.
```
Comment (leoggonzalez) [inline README.md:9]: Shimmer includes a suite of styled components that can be implemented in react and also in plain html, or slim.

There's no Variable component

File: README.md [hunk 2]
```diff
@@ -1,8 +1,26 @@
...
+
+#### Stack Component
+
+The Stack Component, a flexible and reusable component based on the CSS Grid, allows you to easily manage the layout of your app. It is an invaluable tool for creating responsive grid layouts and managing the spacing between items in a container.
```
Comment (leoggonzalez) [inline README.md:13]: Stack component is based on flexbox not grid

File: README.md [hunk 3]
```diff
@@ -1,8 +1,26 @@
...
+
+For detailed usage instructions and examples, refer to the [Component Styles README](./components/styles/README.md) in the `components/styles/stack` folder.
+
+#### Variables Component
```
Comment (leoggonzalez) [inline README.md:17]: not a component

File: components/styles/readme.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
+Stack Component
+===============
+
+The Stack Component is a powerful, reusable SCSS tool built upon the robust CSS Grid framework. It's designed to manage both layout and spacing, thus promoting uniformity, readability, and maintainability within your CSS code.
```
Comment (leoggonzalez) [inline components/styles/readme.md:4]: it's not scss tool, it's a react component and not built with grid, rather with flexbox

File: components/styles/readme.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+Key Features
+------------
+
+*   **Streamlined Grid Layouts**: Create responsive grid layouts that adapt across devices.
```
Comment (leoggonzalez) [inline components/styles/readme.md:9]: not grid

File: components/styles/readme.md [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+*   **Uniform Spacing**: Achieve consistent spacing between items within a container with minimal effort.
+*   **Customizable to Your Needs**: Fine-tune the Stack Component to meet your project's unique design requirements.
+
+Usage
+-----
+
+To use the Stack Component, you need to import it into your SCSS file:
```
Comment (leoggonzalez) [inline components/styles/readme.md:16]: You either import the React component or the full index.css and refer to the classes

**Review by leoggonzalez — DISMISSED** [review]

File: components/stack.tsx [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        gapWidescreen && `stack--widescreen-${gapWidescreen}`,
+        align && `stack--align-${align}`,
+        alignTablet && `stack--align-tablet-${alignTablet}`,
+        alignDesktop && `stack--align-desktop-${alignDesktop}`,
+        alignWidescreen && `stack--align-widescreen-${alignWidescreen}`,
+        justify && `stack--justify-${justify}`,
+        justifyTablet && `stack--justify-tablet-${justifyTablet}`,
+        justifyDesktop && `stack--justify-desktop-${justifyDesktop}`,
+        justifyWidescreen && `stack--justify-widescreen-${justifyWidescreen}`,
+        { "stack--line": line },
```
[the remark this answers is not in the document]
Comment (grcldq) [reply components/stack.tsx]: fixed

File: README.md [hunk 8]
```diff
@@ -4,6 +4,43 @@ Shimmer is a collection of Rails extensions that bring advanced UI features into
...
+import { Stack } from '@nerdgeschoss/shimmer/components/stack';
+
+<Stack gapTablet={4} gapDesktop={12} line>
+  <div></div>
+  <div></div>
+  <div></div>
+</Stack>
+```
+
+To use it in an HTML file, you can just import the css file directly from `@nerdgeschoss/shimmer/components/stack.css` and just implement the classes as they are in the stylesheet:
```
Comment (jmarsh24) [inline README.md:27]:
```suggestion
To use it in an HTML file, you can just import the css file directly from `@nerdgeschoss/shimmer/components/stack.scss` and just implement the classes as they are in the stylesheet:
```

File: README.md [hunk 9]
```diff
@@ -4,6 +4,43 @@ Shimmer is a collection of Rails extensions that bring advanced UI features into
...
+
+```html
+<div class="stack stack--line stack--tablet-4 stack--desktop-12">
```
Comment (jmarsh24) [inline README.md:30]: Would it be good to also explain that we can also change from vertical or horizontal based upon the breakpoints?

File: README.md [hunk 10]
```diff
@@ -4,6 +4,43 @@ Shimmer is a collection of Rails extensions that bring advanced UI features into
...
+  <div></div>
+  <div></div>
+  <div></div>
+</div>
+```
+
```
Comment (jmarsh24) [inline README.md:36]: Should we also explain the different sizes (numbers that are supported?)

File: README.md (hunk 8 above)
Comment (leoggonzalez) [reply README.md:27]: There's also a css export, right? @grcldq

**Review by jmarsh24 — COMMENTED** [review]
I know it's a bit late in the sprint, but it would be super nice that the image we use in the readme could actually be pretty.

File: README.md (hunk 8 above)
Comment (grcldq) [reply README.md:27]: Once it's compiled it will be a .css file, not scss

File: README.md (hunk 9 above)
Comment (grcldq) [reply README.md:30]: Actually I did put it briefly on the longer paragraph, but I was thinking if we want to be more in depth here, should we list out the props and what they are?

File: README.md (hunk 10 above)
Comment (grcldq) [reply README.md:36]: Same here as for the comment above

**Comment by grcldq** [comment]
> I know it's a bit late in the sprint, but it would be super nice that the image we use in the readme could actually be pretty.

I can update it to look better tomorrow for the readme :) for this one I just took a screenshot of how it currently looks like in the styleguide slim file.

File: README.md (hunk 9 above)
Comment (leoggonzalez) [reply README.md:30]: you can do a table with a list for the props like i did here

https://github.com/nerdgeschoss/react-use-form-library#api

File: package.json [hunk 11]
```diff
@@ -4,39 +4,56 @@
...
+    "Geraldine Atayan"
   ],
   "license": "MIT",
   "dependencies": {
     "@hotwired/stimulus": "^3.0.1",
     "@popperjs/core": "^2.11.0",
-    "@rails/request.js": "^0.0.6"
+    "@rails/request.js": "^0.0.6",
+    "classnames": "^2.3.2",
+    "react": "^17.0.1",
```
Comment (JensRavens) [inline package.json:31]: those should be peer dependencies. I don't want to include react in my rails apps.

File: rollup.config.js [hunk 12]
```diff
@@ -12,15 +12,24 @@ export default {
     esbuild({
       target: "es6",
     }),
+    [
```
Comment (JensRavens) [inline rollup.config.js:15]: nothing changed about the rails dependency, so there shouldn't be changes in this file

File: stack.png
Comment (JensRavens) [inline stack.png:1]: default no props does not match the acceptance criteria (it should be full width)

File: README.md (hunk 9 above)
Comment (grcldq) [reply README.md:30]: fixed by @acidburnt

File: package.json (hunk 11 above)
Comment (grcldq) [reply package.json]: moved react and react-dom to peer dependencies

File: rollup.config.js (hunk 12 above)
Comment (grcldq) [reply rollup.config.js]: fixed

File: stack.png
Comment (grcldq) [reply stack.png:1]: fixed, it was an issue with setting the width/height specifically, so I just put in text content instead to be able to output the layout display we want to see

**Review by JensRavens — APPROVED** [review]

**Pull request merged**